dns_server.go

```diff
@@ -19,129 +19,106 @@ import (
-	if err != nil {
-		return ""
+	for _, addr := range addresses {
+		if addr.Is6() {
+			qType := dns.TypeToString[q.Qtype]
+			rr, err := dns.NewRR(fmt.Sprintf("%s %s %s", name, qType, addr.String()))
+			if err == nil {
+				d.dnsMap[q] = append(d.dnsMap[q], rr)
+				d.l.Debugf("DNS record added %s", rr.String())
+			}
+		}
 	}
-	return string(b)
 }
 
-// Add adds the first IPv4 and IPv6 address that appears in `addresses` as the record for `host`
-func (d *dnsRecords) Add(host string, addresses []netip.Addr) {
-	host = strings.ToLower(host)
+func (d *dnsRecords) AddPTR(name string, addresses []netip.Addr) {
 	d.Lock()
 	defer d.Unlock()
-	haveV4 := false
-	haveV6 := false
+
 	for _, addr := range addresses {
-		if addr.Is4() && !haveV4 {
-			d.dnsMap4[host] = addr
-			haveV4 = true
-		} else if addr.Is6() && !haveV6 {
-			d.dnsMap6[host] = addr
-			haveV6 = true
-		}
-		if haveV4 && haveV6 {
-			break
+		arpa, err := dns.ReverseAddr(addr.String())
+		if err == nil {
+			q := dns.Question{Name: arpa, Qclass: dns.ClassINET, Qtype: dns.TypePTR}
+			qType := dns.TypeToString[q.Qtype]
+			rr, err := dns.NewRR(fmt.Sprintf("%s %s %s", arpa, qType, name))
+			if err == nil {
+				d.dnsMap[q] = []dns.RR{rr}
+				d.l.Debugf("DNS record added %s", rr.String())
+			}
 		}
 	}
 }
 
-func (d *dnsRecords) isSelfNebulaOrLocalhost(addr string) bool {
-	a, _, _ := net.SplitHostPort(addr)
-	b, err := netip.ParseAddr(a)
-	if err != nil {
-		return false
-	}
+func (d *dnsRecords) AddTXT(name string, crt cert.Certificate) {
+	d.Lock()
+	defer d.Unlock()
+	q := dns.Question{Name: name, Qclass: dns.ClassINET, Qtype: dns.TypeTXT}
+	d.dnsMap[q] = nil
 
-	if b.IsLoopback() {
-		return true
+	qType := dns.TypeToString[q.Qtype]
+	rr, err := dns.NewRR(fmt.Sprintf("%s %s \"Name: %v\" \"Networks: %v\" \"Groups: %v\" \"UnsafeNetworks: %v\"", name, qType, crt.Name(), crt.Networks(), crt.Groups(), crt.UnsafeNetworks()))
+	if err == nil {
+		d.dnsMap[q] = []dns.RR{rr}
+		d.l.Debugf("DNS record added %s", rr.String())
 	}
+}
 
-	//if we found it in this table, it's good
-	return d.myVpnAddrsTable.Contains(b)
+func (d *dnsRecords) Add(crt cert.Certificate, addresses []netip.Addr) {
+	host := dns.Fqdn(strings.ToLower(crt.Name() + dnsSuffix))
+	d.AddA(host, addresses)
+	d.AddAAAA(host, addresses)
+	d.AddPTR(host, addresses)
+	d.AddTXT(host, crt)
 }
 
-func (d *dnsRecords) parseQuery(m *dns.Msg, w dns.ResponseWriter) {
+func (d *dnsRecords) parseQuery(m *dns.Msg) {
 	for _, q := range m.Question {
 		switch q.Qtype {
-		case dns.TypeA, dns.TypeAAAA:
-			qType := dns.TypeToString[q.Qtype]
-			d.l.Debugf("Query for %s %s", qType, q.Name)
-			ip := d.Query(q.Qtype, q.Name)
-			if ip.IsValid() {
-				rr, err := dns.NewRR(fmt.Sprintf("%s %s %s", q.Name, qType, ip))
-				if err == nil {
-					m.Answer = append(m.Answer, rr)
-				}
-			}
-		case dns.TypeTXT:
-			// We only answer these queries from nebula nodes or localhost
-			if !d.isSelfNebulaOrLocalhost(w.RemoteAddr().String()) {
```

Comment:
I'll admit I don't know the full history on why we had this check, but I hesitate to remove it. I don't want to encourage people to obtain information about a Nebula host (like group membership) _outside_ of a Nebula tunnel.